internal/pkg/propagation/summaries.go

```diff
@@ -0,0 +1,761 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package propagation
+
+// A summary captures the behavior of a function with respect to taint
+// propagation. Specifically, given that at least one of the necessary
+// arguments is tainted, which arguments/return values become tainted?
+// Note that when it's present, the receiver counts as an argument.
+//
+// As an example, consider fmt.Fprintf:
+//   func Fprintf(w io.Writer, format string, a ...interface{}) (n int, err error) {
+// Its summary is:
+//   "fmt.Fprintf": {
+//   	ifTainted:   0b110,
+//   	taintedArgs: []int{0},
+//   },
+// In English, this says that if the format string or the varargs slice are
+// tainted, then the Writer is tainted.
+type summary struct {
+	// ifTainted is a bitset which contains positions for parameters
+	// such that if one of these parameters is tainted, taint should
+	// be propagated to the arguments and return values.
+	// There is a 1-to-1 mapping between the bits and the function's
+	// parameters, with the least significant bit corresponding to the
+	// first (0th) argument.
+	ifTainted int64
+	// the positions of the arguments that taint propagates to if one of the
+	// positions in ifTainted is tainted
+	taintedArgs []int
+	// the positions of the return values that taint propagates to if one of the
+	// positions in ifTainted is tainted
+	taintedRets []int
+}
+
+var fromFirstArgToFirstRet = summary{
+	ifTainted:   0b1,
+	taintedRets: []int{0},
+}
+
+// funcSummaries contains summaries for regular functions
+// that could be called statically.
+var funcSummaries = map[string]summary{
```

Comment:
Some nits throughout:

* Consider dropping the comment trailing `{`.
* Spacing is inconsistent in some interface comments, sometimes even within a single comment, e.g.
```golang
	// type context.Context interface {
	//  Err() error
	//	Value(key interface{}) interface{}
	// }
```

It smells like your IDE swapped some tabs for spaces.

Reply:
* What do you mean by "dropping the comment trailing `{`" ?
* Interesting. Fixed, put spaces everywhere. Side question: how did you detect that? I don't keep any visual cues for tabs/spaces so I had no idea.

Reply:
> What do you mean by "dropping the comment trailing {" ?
Sorry, I meant in the comment strings prior to each summary definition.  You've copy-pasted the function signature lines, which includes a trailing `{`, e.g.,

```golang
	// func Fprint(w io.Writer, a ...interface{}) (n int, err error) {
```

The trailing `{` makes me think that there should be some additional information within the comment.  The function signature is just `func Fprintf(w io.Writer, a ...interface{}) (n int, err error)` - no trailing `{`.

> how did you detect that?

Good ol' fashion eye-ballin'.  It was noticeable in the GitHub diff, even though it's not apparent in my IntelliJ view either.

Reply:
Ah, gotcha. Done.